thrall/app/lib/RetryHandler.scala

```diff
@@ -0,0 +1,40 @@
+package lib
+
+import akka.actor.ActorSystem
+import akka.pattern.{after, retry}
+import play.api.{Logger, MarkerContext}
+
+import scala.concurrent.{ExecutionContext, Future, TimeoutException}
+import scala.concurrent.duration.FiniteDuration
+
+object RetryHandler {
+  def handleWithTimeout[T](f: () => Future[T], attemptTimeout: FiniteDuration)(implicit actorSystem: ActorSystem, executionContext: ExecutionContext,mc: MarkerContext): () => Future[T] = () => {
+
+    val timeout = after(attemptTimeout, using = actorSystem.scheduler)(Future.failed(
+      (() => {
+        Logger.warn(s"Timeout of $attemptTimeout reached.")
+        new TimeoutException(s"Timeout of $attemptTimeout reached.")
+      })()
+    ))
+    Future.firstCompletedOf(Seq(timeout, f()))
+  }
+
+  def handleWithRetry[T](f: () => Future[T], retries: Int, delay: FiniteDuration)(implicit actorSystem: ActorSystem, executionContext: ExecutionContext,mc: MarkerContext): () => Future[T] = () => {
+    implicit val scheduler = actorSystem.scheduler
+    var count = 0
+    def attempt = () => {
+      if (count > 0 ) {
+        Logger.info(s"Attempt $count of $retries")
+      } else {}
+      count = count + 1
+      f()
+    }
+    retry(attempt, retries, delay)
+  }
+
+  def handleWithRetryAndTimeout[T](f: () => Future[T], retries: Int, timeout: FiniteDuration)(implicit actorSystem: ActorSystem, executionContext: ExecutionContext, mc: MarkerContext): () => Future[T] = {
```

Comment:
After going over the code with @AWare, the logic here makes sense given the kinds of failures we are expecting, but it'd be good to document the strategy we are using here, and why we're using it, for other developers.